actionview/lib/action_view/template.rb

```diff
@@ -237,39 +237,40 @@ def encode!
       end
     end
 
-    protected
+    # Templates can be used concurrently in threaded environments
+    # so compilation and any instance variable modification must
+    # be synchronized
+    def thread_safe_compile!(view)
+      return if @compiled
 
-      # Compile a template. This method ensures a template is compiled
-      # just once and removes the source after it is compiled.
-      def compile!(view) #:nodoc:
+      @compile_mutex.synchronize do
+        # Any thread holding this lock will be compiling the template needed
+        # by the threads waiting. So re-check the @compiled flag to avoid
+        # re-compilation
         return if @compiled
+        compile!(view)
+      end
+    end
 
-        # Templates can be used concurrently in threaded environments
-        # so compilation and any instance variable modification must
-        # be synchronized
-        @compile_mutex.synchronize do
-          # Any thread holding this lock will be compiling the template needed
-          # by the threads waiting. So re-check the @compiled flag to avoid
-          # re-compilation
-          return if @compiled
-
-          if view.is_a?(ActionView::CompiledTemplates)
-            mod = ActionView::CompiledTemplates
-          else
-            mod = view.singleton_class
-          end
-
-          instrument("!compile_template") do
-            compile(mod)
-          end
+    # Compile a template. This method ensures a template is compiled
+    # just once and removes the source after it is compiled.
+    def compile!(view = nil) #:nodoc:
```

Comment:
@Mession if you want to you can change `compile!` because code outside of Rails itself shouldn't be using it. "nodoc" means it's a "private" method - meaning it can be changed or deprecated at anytime without a deprecation cycle. Only the framework should be using this method.